x/valset-pref/simulation/sim_msgs.go

```diff
@@ -31,88 +31,123 @@ func RandomMsgSetValSetPreference(k valsetkeeper.Keeper, sim *osmosimtypes.SimCt
-	// check if the delegator valset created
-	err := GetRandomExistingValSet(ctx, k, sim, delegator.Address)
-	if err != nil {
-		return nil, err
-	}
+	delAddr := delegator.Address
 
-	// check that the delegator has delegated tokens
-	_, err = GetRandomExistingDelegation(ctx, k, sim, delegator.Address)
+	// existing delegations
+	delegations, err := k.GetDelegationPreferences(ctx, delAddr.String())
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("no delegations found")
 	}
 
-	remainingWeight := sdk.NewDec(1)
-	preferences, err := GetRandomValAndWeights(ctx, k, sim, remainingWeight)
-	if err != nil {
-		return nil, err
-	}
+	for _, dels := range delegations.Preferences {
+		val, err := sdk.ValAddressFromBech32(dels.ValOperAddress)
+		if err != nil {
+			return nil, fmt.Errorf("validator address not formatted")
+		}
 
-	return &types.MsgRedelegateValidatorSet{
-		Delegator:   delegator.Address.String(),
-		Preferences: preferences,
-	}, nil
-}
+		if sim.StakingKeeper().HasReceivingRedelegation(ctx, delAddr, val) {
+			return nil, fmt.Errorf("receveing redelegation is not allowed for source validators")
+		}
 
-func RandomMsgWithdrawRewardsFromValSet(k valsetkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) (*types.MsgWithdrawDelegationRewards, error) {
-	delegator := sim.RandomSimAccount()
+		if sim.StakingKeeper().HasMaxUnbondingDelegationEntries(ctx, delAddr, val) {
+			return nil, fmt.Errorf("keeper does have a max unbonding delegation entries")
+		}
 
-	err := GetRandomExistingValSet(ctx, k, sim, delegator.Address)
-	if err != nil {
-		return nil, err
+		// check if the user has delegated tokens to the valset
+		_, found := sim.StakingKeeper().GetDelegation(ctx, delAddr, val)
+		if !found {
+			return nil, fmt.Errorf("user hasn't delegated tokens to the validator, %s", val.String())
+		}
 	}
 
-	// check that the delegator has delegated tokens
-	delegations, err := GetRandomExistingDelegation(ctx, k, sim, delegator.Address)
+	// new delegations to redelegate to
+	remainingWeight := sdk.NewDec(1)
+	preferences, err := GetRandomValAndWeights(ctx, k, sim, remainingWeight)
 	if err != nil {
 		return nil, err
 	}
 
-	delegation := delegations[rand.Intn(len(delegations))]
-	validator := sim.StakingKeeper().Validator(ctx, delegation.GetValidatorAddr())
-	if validator == nil {
-		return nil, fmt.Errorf("validator not found")
+	// check if redelegation is possible to new validators
+	for _, vals := range preferences {
+		val, err := sdk.ValAddressFromBech32(vals.ValOperAddress)
+		if err != nil {
+			return nil, fmt.Errorf("validator address not formatted")
+		}
+
+		if sim.StakingKeeper().HasMaxUnbondingDelegationEntries(ctx, delAddr, val) {
+			return nil, fmt.Errorf("keeper does have a max unbonding delegation entries")
+		}
+
+		if sim.StakingKeeper().HasReceivingRedelegation(ctx, delAddr, val) {
+			return nil, fmt.Errorf("receveing redelegation is not allowed for target validators")
+		}
```

Comment:
Why are they not allowed to receive redelegations?

Reply:
because this is the clause in redelegation. For ex: 
ValA --- redelegation --> ValB 
Val --redelegation --> ValC (ERROR: because ValB is receiving redelegation)